### ISSUE TITLE
Questions about code [discussion]

### ISSUE THREAD

**Issue by csBlueChip** [chipweinberger/xesp-usbh#1]
Hi Chip,

You don't seem to have "discussions" enabled so excuse me posting a 'discussion' to the 'issues' tab.

First off: What delightfully easy to read code. Nice job :) ...Your code looks like what I have spent the last couple of weeks trying to find! ...Notably MIDI USB Host for a 'cheap' μC.

1. I see your code has the MIDI Out Endpoint lines commented-out? Is this simply because you have no use for them in your test case, or is this an ESP32-S2 hardware limitation? (EG. the RP2040/Pico which (apparently) only has ONE Bulk Endpoint!)<br/>IE. If I 'uncomment' the lines is it reasonable to expect both MIDI In *and* Out Bulk Endpoints to work?

2. Am I correct to think that the `xesp_*.[c|h]` files are from the ESP IDF? ...Or are these also your work?
 
3. Many MIDI devices (including my CH-345 USB<->MIDI cable) have Descriptors which lie about the size of their Endpoint (8 > 4) ...Do you know where I would need to apply [this (Linux kernel) patch [line 1317]](https://docs.huihoo.com/doxygen/linux/kernel/3.7/sound_2usb_2midi_8c_source.html) in the ESP USB library?

Thanks,
BlueChip

**Comment by chipweinberger** [comment]
thanks for the compliment!

few things:

- yes, only commented out since i didn't implement it. It would work but you'd also need to write a simple transfer out function to actually send data to the endpoint. similar to the transfer in code.
- xesp* is my code. I use that abbreviation since USBH would usually be system library (esp), but this is 3rd party (me) so i added an X
- Espressif is adding official USBH support in esp-df 4.4 which is around the corner, so I would wait for that!!
- this code was built against a esp idf 4.4 commit from ~may 2021 and it appears to not compile against the newest commits so it would need to be updated a bit

in other words, wait for the official support :) 

This code was written so I didn't have to wait around! and we didn't know how long it would be until the official code came out.

**Comment by chipweinberger** [comment]
as for #3 that's the first i'm hearing about that. I'll need to look into that! 

also, what are you working on out of curiosity?

**Comment by csBlueChip** [comment]
/ALL/ of that is yours?! :o ...I think I need to upgrade my level of impressed ;)

I have written (but, obvs, not tested) the code to be a:
* Virtual Keyboard
* Synth
* Router

You can find my [1st draft here](https://github.com/csBlueChip/tinyusb/tree/master/_MIDIH.tmp), there's a block diagram of the I/O side of things in the `readme`
The `synth` files contain the functions you need to write to implement your own synth module ...I'll probably knock up an FM synth as a demo, but that is not a priority.

`cb.c` is the user callbacks - when incoming MIDI events are received
`core.c` is the driver that dispatches callbacks, etc.
It's pretty well commented, so if you want a quick idea of the goal - the API functions are listed at the bottom of `midih.h` and `midiu.h`

I've also got a MID/RIFF player working nicely for format-0 and format-1 (non-SMPTE) ...I got distracted and am currently designing a Linux Console UI (!ncurses) for it &lt;shrug> ...I'll probably push that to a new repo sometime over the next week or two - I keep finding obscure edge-cases :/

Hmmm. Espressif have been promising USBH support for the S2 for over a year now ...I'm cautious to hold my breath any longer on this - especially if I've just found some tried-and-tested code, with an easy to read test harness, that does the exact thing I need.

**Comment by csBlueChip** [comment]
> as for #3 that's the first i'm hearing about that. I'll need to look into that!

I've not tried this cable with Windows in years, but when I first got it, I was getting the *weirdest* problems, SOME events were making it through, but not others ...I jumped through a lot of forums (or was it BBS'es) to find the root cause of the problem - and (anecdotally) it certainly fixed my issues &lt;shrug>

**Comment by chipweinberger** [comment]
Espressif released their USB host code about a month ago.

It's not yet in an official release though so it could be subject to change. But it should be better tested then mine.

If you do use my code, make sure to use a esp-idf commit from around april/may 2021 since it does not work or even compile anymore on newer commits.

It's on my todo list to update this code, but I won't get to that for awhile.
